### PR TITLE
bpftool: add support for split BTF to gen min_core_btf

### PR DESCRIPTION
**Pull request by brycekahle**

### CODE REVIEW

File: src/gen.c
Comment (qmonnet) [inline src/gen.c:2142]: Nit: Let's keep the declarations in this block in reverse-Christmas tree order (longest first) if possible, please.

**Review by qmonnet — CHANGES_REQUESTED** [review]
Hi, and thanks for the patch!

Can you please motivate the change in your commit description? I'm not super familiar with split BTF and minified BTF generation, but unless I'm wrong we need to pass the base BTF on the command line for this to work, don't we? If so, I'd like to have:

- An example invocation in the commit description, to help with the review
- Ideally, an example in the relevant docs (`Documentation/bpftool-gen.rst`)
- An update of the docs if option `-B` is required for this to work (so far we only document `-B` in the `bpftool btf` man page), along with the update of the help string in gen.c's `do_help()`

Then I won't merge the commit from this repo, sorry. Instead, you need to submit it to the BPF mailing list (see [the relevant docs](https://docs.kernel.org/bpf/bpf_devel_QA.html#id3) - I can help if necessary). Please be sure to CC the BPF maintainers and myself, and to prefix your commit title with `bpftool: `.

**Comment by brycekahle** [comment]
Updated, let me know if it looks good for me to submit to the mailing list.

File: docs/bpftool-gen.rst
Comment (qmonnet) [inline docs/bpftool-gen.rst:223]: This paragraph would need to be adjusted, it's written with `bpftool btf dump` in mind. From what I understand, if we don't pass `-B` for `bpftool gen min_core_btf`, we just don't use split BTF, right? There's no attempt to guess the path. So this paragraph can maybe just be deleted.

**Review by qmonnet — CHANGES_REQUESTED** [review]
Looks better, thanks! A few more nits:

- Please wrap your commit description (except for the example invocations) on ~72 char-wide lines
- The second paragraph for the description of the `-B` option is probably not required for `bpftool gen`, see below.

After that it looks good from my side, although I've not tested the changes.

I forgot to say it because it seemed obvious to me, but just to be sure, to submit to the BPF mailing list you'll have to port your changes to the Linux git repo (the bpf-next tree). Bpftool sources are under `tools/bpf/bpftool/`.

File: src/gen.c
Comment (qmonnet) [inline src/gen.c:1634]: I'm just seeing you need to update the line above, here, to remove the ending brace. In other words, instead of this:

```
		"                    {-L|--use-loader} }\n"
		"                    {-B|--base-btf} }\n"
```

Having this:

```
		"                    {-L|--use-loader}\n"
		"                    {-B|--base-btf} }\n"
```

**Review by qmonnet — CHANGES_REQUESTED** [review]
Reading again the description, I'd recommend adding this note:

> A minimized module BTF will still not contain vmlinux BTF types, so you should always minimize the vmlinux file first, and then minimize the kernel module file.

to either the option description, or to your example, in the man page. I suspect that, otherwise, users will try to create a the minimised object in one step by passing the path of both the vmlinux and module BTF paths.

By the way, with your patch, you currently end up with two generated minimised BTF objects, `mod-min.btf` and `vm-min.btf`, right? Wouldn't it make sense to look into passing all BTF objects (vmlinux and modules) in one step, and generate a single minimised BTF object from that? Or does it defeat what you're trying to achieve? (Sorry for failing to raise this on the previous review.)

Also, one other nit below.

File: src/gen.c
Comment (brycekahle) [reply src/gen.c:1634]: good catch

**Comment by brycekahle** [comment]
> Wouldn't it make sense to look into passing all BTF objects (vmlinux and modules) in one step, and generate a single minimised BTF object from that? Or does it defeat what you're trying to achieve? (Sorry for failing to raise this on the previous review.)

There may be multiple kernel modules in play for a single eBPF object file, since it can contain multiple programs and they each may probe different modules. Since a kernel module BTF has conflicting type IDs (they all start at max(vmlinux) + 1) with other kernel module BTF, I'm not sure there is a way with libbpf to load them all at the same time.

**Comment by brycekahle** [comment]
Updated

File: docs/bpftool-gen.rst
Comment (qmonnet) [inline docs/bpftool-gen.rst:460]: Please wrap

**Review by qmonnet — APPROVED** [review]
> There may be multiple kernel modules in play...

Might be worth mentioning in your commit description. I don't have a suggestion to address this, but others on the BPF ML may have ideas.

Otherwise, it looks nearly all good from my side (with a last nit below).

Looking forwards to seeing your patch on the ML :)

**Comment by brycekahle** [comment]
Welp, first ML attempt was a failure. Gmail borked the formatting. I'm looking into my options, but my work email is pretty locked down so I cannot even get an app password.

**Comment by qmonnet** [comment]
Yeah GMail interface is known to break patch formatting. I'm using `git send-email`, but yes, it needs an app password to use the SMTP.

Have you looked into using [a local client](https://www.kernel.org/doc/html/latest/process/email-clients.html) that supports 0Auth2 to connect to GMail (I think Thunderbird at least can)? I don't think you'd need an app password to do that?

If you don't find a solution for your work email, you can always use a personal email to send your patch. `git send-email` should preserve the `From:` header and the `Signed-off-by`, so only your work email would appear in the git logs (your personal email would be on https://lore.kernel.org/bpf/ though).

If none of these work or is convenient, I'm happy to post the patch on your behalf.

**Comment by brycekahle** [comment]
Closing because it is on ML now

**Comment by qmonnet** [comment]
https://lore.kernel.org/r/20240130230510.791-1-git@brycekahle.com

**Pull request closed without merging**
